apps/ejabberd/src/mod_register.erl

```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
-					id = ID,
-					sub_el = [SubEl]},
-			    ejabberd_router:route(
-			      jlib:make_jid(User, Server, Resource),
-			      jlib:make_jid(User, Server, Resource),
-			      jlib:iq_to_xml(ResIQ)),
-			    ejabberd_auth:remove_user(User, Server),
-			    ignore;
-			_ ->
-			    IQ#iq{type = error,
-				  sub_el = [SubEl, ?ERR_NOT_ALLOWED]}
-		    end;
-		(UTag /= false) and (PTag /= false) ->
-		    User = xml:get_tag_cdata(UTag),
-		    Password = xml:get_tag_cdata(PTag),
-		    try_register_or_set_password(
-		      User, Server, Password, From,
-		      IQ, SubEl, Source, Lang);
-		true ->
-		    IQ#iq{type = error,
-			  sub_el = [SubEl, ?ERR_BAD_REQUEST]}
-	    end;
-	get ->
-	    {_IsRegistered, UsernameSubels, QuerySubels} =
-		case From of
-		    #jid{user = User, lserver = Server} ->
-			case ejabberd_auth:is_user_exists(User, Server) of
-			    true ->
-				{true, [#xmlcdata{content = User}],
-				 [#xmlel{name = <<"registered">>}]};
-			    false ->
-				{false, [#xmlcdata{content = User}], []}
-			end;
-		    _ ->
-			{false, [], []}
-		end,
-        TranslatedMsg = translate:translate(
-                          Lang, <<"Choose a username and password to register with this server">>),
-		    IQ#iq{type = result,
-		  sub_el = [#xmlel{name = <<"query">>,
-				   attrs = [{<<"xmlns">>, <<"jabber:iq:register">>}],
-				   children = [#xmlel{name = <<"instructions">>,
-				              children = [#xmlcdata{content = TranslatedMsg}]},
-				       #xmlel{name = <<"username">>,
-					      children = UsernameSubels},
-				       #xmlel{name = <<"password">>}
-				       | QuerySubels]}]}
+%% Clients must register before being able to authenticate.
+process_unauthenticated_iq(Sender, Receiver, #iq{type = set} = Stanza, IPAddr) ->
+    process_iq_set(Sender, Receiver, Stanza, IPAddr);
+process_unauthenticated_iq(Sender, Receiver, #iq{type = get} = Stanza, IPAddr) ->
+    process_iq_get(Sender, Receiver, Stanza, IPAddr).
+
+process_iq(From, To, #iq{type = set} = IQ) ->
+    process_iq_set(From, To, IQ, jlib:jid_tolower(From));
+process_iq(From, To, #iq{type = get} = IQ) ->
+    process_iq_get(From, To, IQ, jlib:jid_tolower(From)).
+
+process_iq_set(From, To, #iq{sub_el = Child} = IQ, Source) ->
+    true = is_query_element(Child),
+    handle_set(IQ, From, To, Source).
+
+handle_set(Stanza, ClientJID, ServerJID, Source) ->
+    #iq{sub_el = Query} = Stanza,
+    case has_only_remove_child(Query) of
+        true ->
+            attempt_cancelation(ClientJID, ServerJID, Stanza);
+        {false, more} ->
+            error_response(Stanza, ?ERR_BAD_REQUEST);
+        {false, absent} ->
+            case has_username_and_password_children(Query) of
+                true ->
+                    Credentials = get_username_and_password_values(Query),
+                    register_or_change_password(Credentials, ClientJID, ServerJID, Stanza, Source);
+                false ->
+                    ignore
```

Comment:
should it be `error_response(..., ?ERR_BAD_REQUEST)`?  The same error as several lines above - L120. Both paths are result of invalid user input, but one of them returns error while the other one just ignores iq.